File: android/app/src/main/java/com/zillit/lcw/util/DateUtils.kt
package com.zillit.lcw.util

import java.text.SimpleDateFormat
import java.util.*

/**
 * DateUtils — convert between epoch milliseconds and formatted date strings.
 * All Box Schedule dates are stored as epoch ms (Long).
 */
object DateUtils {

    private val dateFormat = SimpleDateFormat("MMM d, yyyy", Locale.US)
    private val timeFormat = SimpleDateFormat("h:mm a", Locale.US)
    private val dayNameFormat = SimpleDateFormat("EEEE", Locale.US)
    private val monthYearFormat = SimpleDateFormat("MMMM yyyy", Locale.US)
    private val shortDateFormat = SimpleDateFormat("MMM d", Locale.US)
    private val dayMonthFormat = SimpleDateFormat("EEE, MMM d", Locale.US)
    private val fullDayFormat = SimpleDateFormat("EEEE, MMM d, yyyy", Locale.US)

    fun fromEpoch(ms: Long): Date = Date(ms)

    fun toEpoch(date: Date): Long = date.time

    fun formatDate(ms: Long): String = dateFormat.format(Date(ms))

    fun formatTime(ms: Long): String = timeFormat.format(Date(ms))

    fun formatDayName(ms: Long): String = dayNameFormat.format(Date(ms))

    fun formatMonthYear(date: Date): String = monthYearFormat.format(date)

    fun formatShortDate(ms: Long): String = shortDateFormat.format(Date(ms))

    fun formatDayMonth(ms: Long): String = dayMonthFormat.format(Date(ms))

    fun formatFullDay(ms: Long): String = fullDayFormat.format(Date(ms))

    fun startOfDay(date: Date): Date {
        val cal = Calendar.getInstance()
        cal.time = date
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    fun startOfDayMs(ms: Long): Long = startOfDay(Date(ms)).time

    fun isToday(ms: Long): Boolean {
        val today = startOfDay(Date())
        val target = startOfDay(Date(ms))
        return today.time == target.time
    }

    fun isPast(ms: Long): Boolean {
        return startOfDayMs(ms) < startOfDayMs(System.currentTimeMillis())
    }

    fun isWeekend(ms: Long): Boolean {
        val cal = Calendar.getInstance()
        cal.timeInMillis = ms
        val day = cal.get(Calendar.DAY_OF_WEEK)
        return day == Calendar.SATURDAY || day == Calendar.SUNDAY
    }

    fun addDays(ms: Long, days: Int): Long {
        val cal = Calendar.getInstance()
        cal.timeInMillis = ms
        cal.add(Calendar.DAY_OF_MONTH, days)
        return cal.timeInMillis
    }
}

File: ios/ZillitLCW/Utilities/DateUtils.swift
import Foundation

struct DateUtils {
    static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM d, yyyy"
        return f
    }()

    static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "h:mm a"
        return f
    }()

    static let dayNameFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "EEEE"
        return f
    }()

    static let monthYearFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMMM yyyy"
        return f
    }()

    static let shortDateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM d"
        return f
    }()

    static let fullDayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "EEEE, MMM d, yyyy"
        return f
    }()

    static func fromEpoch(_ ms: Int64) -> Date {
        Date(timeIntervalSince1970: TimeInterval(ms) / 1000.0)
    }

    static func toEpoch(_ date: Date) -> Int64 {
        Int64(date.timeIntervalSince1970 * 1000)
    }

    static func formatDate(_ ms: Int64) -> String {
        dateFormatter.string(from: fromEpoch(ms))
    }

    static func formatTime(_ ms: Int64) -> String {
        timeFormatter.string(from: fromEpoch(ms))
    }

    static func formatDayName(_ ms: Int64) -> String {
        dayNameFormatter.string(from: fromEpoch(ms))
    }

    static func formatMonthYear(_ date: Date) -> String {
        monthYearFormatter.string(from: date)
    }

    static func formatShortDate(_ ms: Int64) -> String {
        shortDateFormatter.string(from: fromEpoch(ms))
    }

    static let dayMonthFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "EEE, MMM d"
        return f
    }()

    static func formatDayMonth(_ ms: Int64) -> String {
        dayMonthFormatter.string(from: fromEpoch(ms))
    }

    static func formatFullDay(_ ms: Int64) -> String {
        fullDayFormatter.string(from: fromEpoch(ms))
    }

    static func isToday(_ ms: Int64) -> Bool {
        Calendar.current.isDateInToday(fromEpoch(ms))
    }

    static func isPast(_ ms: Int64) -> Bool {
        fromEpoch(ms) < Calendar.current.startOfDay(for: Date())
    }

    static func isWeekend(_ ms: Int64) -> Bool {
        Calendar.current.isDateInWeekend(fromEpoch(ms))
    }
}
